iroh/src/downloader.rs

```diff
@@ -0,0 +1,979 @@
+//!
+//! The [`Downloader`] interacts with four main components to this end.
+//! - [`Dialer`]: Used to queue opening connections to peers we need to perform downloads.
+//! - [`ProviderMap`]: Where the downloader obtains information about peers that could be
+//!   used to perform a download.
+//! - [`Store`]: Where data is stored.
+//! - [`CollectionParser`]: Used by the Get state machine logic to identify blobs encoding
+//!   collections.
+//!
+//! Once a download request is received, the logic is as follows:
+//! 1. The [`ProviderMap`] is queried for peers. From these peers some are selected
+//!    prioritizing connected peers with lower number of active requests. If no useful peer is
+//!    connected, or useful connected peers have no capacity to perform the request, a connection
+//!    attempt is started using the [`Dialer`].
+//! 2. The download is queued for processing at a later time. Downloads are not performed right
+//!    away. Instead, they are initially delayed to allow the peer to obtain the data itself, and
+//!    to wait for the new connection to be established if necessary.
+//! 3. Once a request is ready to be sent after a delay (initial or for a retry), the preferred
+//!    peer is used if available. The request is now considered active.
+//!
+//! Concurrency is limited in different ways:
+//! - *Total number of active request:* This is a way to prevent a self DoS by overwhelming our own
+//!   bandwidth capacity. This is a best effort heuristic since it doesn't take into account how
+//!   much data we are actually requesting or receiving.
+//! - *Total number of connected peers:* Peer connections are kept for a longer time than they are
+//!   strictly needed since it's likely they will be useful soon again.
+//! - *Requests per peer*: to avoid overwhelming peers with requests, the number of concurrent
+//!   requests to a single peer is also limited.
+
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    num::NonZeroUsize,
+};
+
+use futures::{future::LocalBoxFuture, stream::FuturesUnordered, FutureExt, StreamExt};
+use iroh_bytes::{
+    baomap::{range_collections::RangeSet2, Store},
+    collection::CollectionParser,
+    protocol::RangeSpecSeq,
+    Hash,
+};
+use iroh_net::{key::PublicKey, MagicEndpoint};
+use tokio::sync::{mpsc, oneshot};
+use tokio_util::{sync::CancellationToken, time::delay_queue};
+use tracing::{debug, trace};
+
+mod get;
+mod test;
+
+/// Delay added to a request when it's first received.
+const INITIAL_REQUEST_DELAY: std::time::Duration = std::time::Duration::from_millis(500);
+/// Number of retries initially assigned to a request.
+const INITIAL_RETRY_COUNT: u8 = 4;
+/// Duration for which we keep peers connected after they were last useful to us.
+const IDLE_PEER_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+/// Capacity of the channel used to comunicate between the [`Downloader`] and the [`Service`].
+const SERVICE_CHANNEL_CAPACITY: usize = 120;
```

Comment:
why not a power of two?

Reply:
I think because my brain saw a 128 in the code and then moved it to a constant and 8 and look similar? jk aside I'll change it to 128 again :)

Reply:
> I think because my brain saw a 128 in the code and then moved it to a constant and 8 and 0 look similar? jk aside I'll change it to 128 again :)

🤣 



> fixed in [12c63dd](https://github.com/n0-computer/iroh/commit/12c63dd5a6950a339e23c6a8eaabbc8c666953ed)

thanks, makes my brain happy